models/maat/R/model2netcdf.MAAT.R

```diff
@@ -96,101 +69,145 @@ model2netcdf.MAAT <- function(rundir, outdir, sitelat = -999, sitelon = -999, st
-      ## !!TODO: ADD MORE MAAT OUTPUTS HERE!! ##
-
+      met_tbl <- tibble::tribble(
+        ~data, ~oldname, ~newname, ~oldunits, ~newunits, ~longname,
+        list(out.year), "Year", "Year", "YYYY", "YYYY", "Simulation Year",
+        list(tvals), "FracJulainDay", "FracJulianDay", "Frac DOY", "Frac DOY", "Fraction of Julian Date",
+        )
+      process_tbl <- rbind.data.frame(met_tbl, process_tbl)
     } else {
+      bounds <- array(data = NA, dim = c(1,2))
+      bounds[,1] <- 0
+      bounds[,2] <- 1
       t <- ncdf4::ncdim_def(name = "time", units = paste0("days since ", maat_run_start_date),
                             vals = 1, calendar = "standard", 
                             unlim = TRUE)  # standard calendar for leap years?  Also need to be sure we update cal depending on leap/no leap
-      output <- NULL
-      ncdims <- list(lon, lat, t) 
-      output <- var_update(sub.maat.output$A, output, "A", "assimilation_rate", oldunits="umol C m-2 s-1", newunits="kg C m-2 s-1", missval=-999, 
-                           longname="Leaf assimilation rate", ncdims=ncdims)
-      output <- var_update(maat.output$rd, output, "rd", "leaf_respiration", oldunits="umol C m-2 s-1", newunits="kg C m-2 s-1", missval=-999,
-                           longname="Leaf Respiration Rate", ncdims=ncdims)
-      output <- var_update((1/(maat.output$rs)), output, "gs", "stomatal_conductance", oldunits="mol H2O m-2 s-1",
-                           newunits="kg H2O m-2 s-1", missval=-999, longname="Leaf Stomatal Conductance", ncdims=ncdims)
-      output <- var_update(maat.output$ci, output, "ci", "Ci", oldunits="Pa",
-                           newunits="Pa", missval=-999, longname="Leaf Internal CO2 Concentration", ncdims=ncdims)
-      output <- var_update(maat.output$cc, output, "cc", "Cc", oldunits="Pa",
-                           newunits="Pa", missval=-999, longname="Leaf Mesophyll CO2 Concentration", ncdims=ncdims)
-      ## !!TODO: ADD MORE MAAT OUTPUTS HERE!! ##  
     }
+
+    ncdims <- list(lon, lat, t)
+
+    ## Subset data for processing
+    # setup netCDF time variable for year
+    time_interval <- ncdf4::ncdim_def(name = "hist_interval", 
+                                      longname = "history time interval endpoint dimensions",
+                                      vals = 1:2, units="")
+
+    #output <- purrr::pmap(process_maat_variable, process_tbl, missval = -999, ncdims = ncdims)
+    output <- purrr::pmap(process_tbl, process_maat_variable, missval = -999, ncdims = ncdims)
+
+    output <- c(output, list(
+      var = ncdf4::ncvar_def(name = "time_bounds", units = "", 
+                             longname = "history time interval endpoints", 
+                             dim = list(time_interval, time = t), 
+                             prec = "double"),
+      dat = c(rbind(bounds[, 1], bounds[, 2]))
+    ))
```

Comment:
Found the issue! The `list` call here needs to be wrapped in another `list` to get this to work as expected.

```r
    output <- c(output, list(list(
      var = ncdf4::ncvar_def(name = "time_bounds", units = "", 
                             longname = "history time interval endpoints", 
                             dim = list(time_interval, time = t), 
                             prec = "double"),
      dat = c(rbind(bounds[, 1], bounds[, 2]))
    )))
```

Basically, because of the way `c.list` works, this is currently adding _two_ elements to the `output` list (`var` and `dat`), when it should be adding a _single_ element which is itself a list containing `var` and `dat`. By wrapping the `list` in another `list` call, you are forcing it to be a single element which is a `list` containing two objects inside of it.

Reply:
Still seems like something is not quite right here:
```
    ## write netCDF data
    nc_vars <- purrr::map(output, "var") # Extract var from each output
    ncout <- ncdf4::nc_create(file.path(outdir, paste(year, "nc", sep = ".")), nc_vars)
    on.exit(ncdf4::nc_close(ncout))
    ncdf4::ncatt_put(ncout, "time", "bounds", "time_bounds", prec=NA)
    for (output_var in output) {
      #print(i)  # for debugging
      ncdf4::ncvar_put(ncout, output_var$var, output_var$dat)
    }
```

should that be nc_vars in output?  Also, getting a dimension error writing to the nc file, perhaps the set dims are wrong


```
  ncvar_put: error: you asked to write 17520 values, but the passed data array only has 1 entries!
>     for (output_var in output) {
+       print(output_var$var$name)  # for debugging
+       ncdf4::ncvar_put(ncout, output_var$var, output_var$dat)
+     }
[1] "Year"
Error in ncdf4::ncvar_put(ncout, output_var$var, output_var$dat) :
  ncvar_put: error: you asked to write 17520 values, but the passed data array only has 1 entries!
>
```

Reply:
> output[[1]]$var$ndims
[1] 3

$ndims
[1] 3

$varsize
[1]     1     1 17520

hmmm......

Reply:
OK, another guess at the problem: The `year` (`"Year"`) variable is a constant -- just one value -- but its variable dimensions are set to the number of timesteps. So the solution is to either:

(1) Cycle the `year` variable across the dimensions of the NetCDF file (`year <- rep(year, ncdims[[X]])` where X is whatever the correct dimension is -- I think 3?), or

(2) Set the dimensions of the "Year" variable to 1 (or [1,1,1]).

Reply:
Just kidding. I think maybe the _real_problem is that the `data` being passed into `process_met_tbl` is a length 1 `list`. In my `tibble::tribble` calls, try getting rid of the `list(...)` surrounding each `data` element -- turns out code like this does exactly what I want without wrapping complex elements in `list` (because `x` is automatically determined to be a list column):

```r
a <- 1:10
b <- 2:15
tibble::tribble(
  ~x, ~y,
  a, 5,
  b, 27
)
```

Reply:
@ashiklom I am getting lost in this changes at this point because I havent had uninterrupted attention on this, Can we please put this PR to bed this week so I can move on?  Perhaps we can just figure out what else I need to do to make this work on close it, we can work on further edits later.